safekeeper/src/wal_backup.rs

```diff
@@ -241,7 +242,7 @@ impl WalBackupTask {
 
         let mut retry_attempt = 0u32;
         // offload loop
-        loop {
+        while !self.timeline.cancel.is_cancelled() {
```

Comment:
This will still block on the inner awaits (i.e. the retry sleep and `commit_lsn_watch_rx`). Consider turning the loop body into an async block/function and selecting on it and the cancellation token instead, which will cancel any in-flight operation.

Reply:
Good catch, I think I had mistakenly assumed that the tx side of the channel was getting dropped already during shutdown, but no.

Added proper cancellation handling in fa50815c0

Reply:
Did you forget to remove the `sleep` there? Seems like we're now waiting twice.

There's also a `get_wal_backup_lsn()` lock access that doesn't respect cancellation. Maybe that's ok to ignore -- it should only really matter if there was a deadlock.

Reply:
>Did you forget to remove the sleep there? Seems like we're now waiting twice.

Argh, good spot, thanks.  Fixed in 40c3c114d

>There's also a get_wal_backup_lsn() lock access that doesn't respect cancellation. Maybe that's ok to ignore -- it should only really matter if there's a deadlock.

Right - the design expectation is that once we're shutting down, anything holding the lock should be releasing it promptly (unless, as you say, there were a deadlock)